src/commands/run.rs

```diff
@@ -423,22 +426,46 @@ impl RunCommand {
         // we don't actually read the whole file into memory. Note that this
         // behavior is disabled by default, though, because it's not safe to
         // pass arbitrary user input to this command with `--allow-precompiled`
-        let mut file =
-            File::open(path).with_context(|| format!("failed to open: {}", path.display()))?;
-        let mut magic = [0; 4];
-        if let Ok(()) = file.read_exact(&mut magic) {
-            if &magic == b"\x7fELF" {
-                if self.allow_precompiled {
-                    return unsafe { Module::deserialize_file(engine, path) };
+
+        // Try to mmap the module file.
+        // If it fails(it wouldn't be a regular file), allocate a fixed mmap buffer
+        // to store the whole bytes in a memory.
+        let mmap_vec = MmapVec::from_file(path).or_else(|_| {
+            // Mmap-ing this much would not eat system memory thanks for paging
+            const MAX_READ_FROM_STREAM: usize = 1 << 30; // TODO: make this configurable with environment variables?
```

Comment:
Arbitrary limitations and environment variables are something we'd like to avoid if possible. The users of `mmap_vec` don't care if they have an `MMapVec` or a plain `Vec<u8>`, so in the case where the mmap fails, would it work to just call `std::fs::read` to read the file? That would require reorganizing this code to avoid using `MmapVec` as the type to unify the two options. Something like this:

```rust
    let mmap_vec;
    let read;
    let bytes: &[u8] = if let Ok(mapped) = MmapVec::from_file(Path::new(path)) {
        mmap_vec = mapped;
        mmap_vec.as_ref()
    } else {
        read = std::fs::read("hello.txt")?;
        read.as_ref()
    };
```

Reply:
Yeah I misinterpreted @bjorn3 's first review so I'll roll back into previous implementation which just uses `Read::read_to_end` and `Vec<u8>`, and prevent redundant allocation when a module is from a regular file. This *may* be sub-optimal for irregular file case but I think it's OK because no one wants to do `hyperfine 'cat foo.wasm | wasmtime -'`?